File: src/hilbert/models/session.py
"""Session and state models for Hilbert."""

from datetime import datetime
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, Field


class SessionStatus(str, Enum):
    """Status of a research session."""

    PLANNING = "planning"
    SEARCHING = "searching"
    MERGING = "merging"
    SYNTHESIZING = "synthesizing"
    REVIEWING = "reviewing"
    VERIFYING = "verifying"
    WRITING = "writing"
    DONE = "done"
    ERROR = "error"


class Session(BaseModel):
    """Research session stored in SQLite."""

    session_id: str
    query: str
    max_rounds: int = 3
    current_round: int = 0
    status: SessionStatus = SessionStatus.PLANNING
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    error_message: Optional[str] = None
    tags: List[str] = Field(default_factory=list)

    def is_resumable(self) -> bool:
        """Check if session can be resumed."""
        return self.status not in (SessionStatus.DONE, SessionStatus.ERROR)


class Checkpoint(BaseModel):
    """State checkpoint for session resume."""

    checkpoint_id: int = 0
    session_id: str
    round: int
    state_json: str
    created_at: datetime = Field(default_factory=datetime.now)

File: cli/src/core/models/session.ts
import { z } from "zod";

export const SessionStatusSchema = z.enum([
  "planning",
  "searching",
  "merging",
  "synthesizing",
  "reviewing",
  "verifying",
  "writing",
  "done",
  "error",
]);

export type SessionStatus = z.infer<typeof SessionStatusSchema>;

export const SessionSchema = z.object({
  session_id: z.string(),
  query: z.string(),
  max_rounds: z.number().default(3),
  current_round: z.number().default(0),
  status: SessionStatusSchema.default("planning"),
  created_at: z.string().default(() => new Date().toISOString()),
  updated_at: z.string().default(() => new Date().toISOString()),
  error_message: z.string().optional(),
  tags: z.array(z.string()).default([]),
  last_searched_at: z.string().optional(),
});

export type Session = z.infer<typeof SessionSchema>;

export function isResumable(session: Session): boolean {
  return !["done", "error"].includes(session.status);
}

export const CheckpointSchema = z.object({
  checkpoint_id: z.number().default(0),
  session_id: z.string(),
  round: z.number(),
  state_json: z.string(),
  created_at: z.string().default(() => new Date().toISOString()),
});

export type Checkpoint = z.infer<typeof CheckpointSchema>;